Make a simple functional comment feature

helenatran opened the pull request:
So far, my comment feature consists of:
- a HTML form
- GET request (to get the comments). Comments are then displayed with an unordered list. 
- POST request (to submit the comments)
- Store comments with Datastore

helenatran committed:
Update DataServlet.java

helenatran committed:
Use fetch() to request 'Hello Helena' from server and add it to the page

helenatran committed:
Merge branch 'master' into server

helenatran committed:
Add list of messages on portfolio using fetch

helenatran committed:
Make comments feature functional

helenatran committed:
Update getComments function

helenatran committed:
Add Datastore dependency

helenatran committed:
Store comments using Datastore

g-wittie commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -14,19 +14,63 @@
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
+  
+  ArrayList<String> comments = new ArrayList<String>();
```
Unless there's a good reason (e.g. performance) prefer omitting the specific type for the var as it's an implementation detail, so:
`List<String> comments = new ArrayList<String>();`

This is because the type of list (ArrayList vs LinkedList vs WhateverOtherList) doesn't really matter as long as it's a List (it inherits from List and hence supports the List interface).

I'm not sure if this works on the version of Java you're on (it probably does) but you should be able to omit the type on the right-hand side with an empty diamond operator, like so:
`List<String> comments =  new ArrayList<>();`

g-wittie commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -14,19 +14,63 @@
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    datastore.put(commentEntity);
+
+    String commentDetails = "Username: " + username + "; Comment: " + comment;
```
I think this version of Java already has 'printf', in which case you should use that rathre than concatenating with '+' (see https://www.baeldung.com/java-printstream-printf).

Something like 
`System.out.printf("Username: %s; Comment: %s;", username, comment);`

The underlying reason is that the concat operator ('+') actually copies the entire string each time you concatenate to it. Not a big deal in a simple case like this, but it can get very expensive when dealing with big strings and/or iteratively constructing strings, so it's good to get into the habit of avoiding the '+' operator with strings.

helenatran replied:
I ended up making some changes to my code so I am not using a string like that anymore. But I had a question: from what I understand, it's better to avoid the concat operator with strings in Java. However, I've been using it quite often with JavaScript (and that might be why I found it normal to use it here!). Is it also the case for JavaScript? Or does the reason you have mentioned only applies to Java?

g-wittie commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -14,19 +14,63 @@
+   * Converts the randomList instance into a JSON string using the Gson library.
+   */
+  private String convertToJson(ArrayList<String> list) {
+    Gson gson = new Gson();
```
You might want to create an instance variable for this in the class so it can be reused. Objects like this are often relatively expensive to build, so you want to avoid doing so on every method call.

helenatran replied:
I ended up getting rid of this function and have this line at the beginning of the class; but I am not sure whether that's better or no? Is there any other way I could have done it?

helenatran committed:
Load entities using queries + fix issues

helenatran commented:
Thanks for the review @g-wittie! I pushed a few more changes that I did before seeing the reviews (I will avoid pushing more changes to this branch from now), including:
- Loading entities with queries to get the comments
- Added a timestamp attribute to the comment so that they can be sorted chronologically 
- Added a deleteDataServlet to deal with deleting comment

lorikeet-s commented on portfolio/src/main/webapp/index.html:
```diff
@@ -108,6 +108,16 @@ <h4>Fun Fact Roulette</h4>
         </div>
+        <h4>Leave me a comment: </h4>
+        <form action="/data" method="POST">
+          <label for="username">Name: </label>
```
I think the for attribute is supposed to match the id of another component, so it knows what it's labelling.

lorikeet-s commented on portfolio/src/main/webapp/script.js:
```diff
@@ -106,3 +106,49 @@ function getAboutMe(aboutMeId) {
+async function loadComments() {
+    const response = await fetch('/data');
+    const comments = await response.json();
+    console.log(comments);
```
remove debugging logs before you submit, unless it's intentionally kept for production logging purposes.

lorikeet-s commented on portfolio/src/main/java/com/google/sps/servlets/DeleteDataServlet.java:
```diff
@@ -0,0 +1,39 @@
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    datastore.delete(commentEntityKey);
+  }
+}
```
add new line

lorikeet-s commented on portfolio/src/main/webapp/script.js:
```diff
@@ -106,3 +106,49 @@ function getAboutMe(aboutMeId) {
+  const params = new URLSearchParams();
+  params.append('id', comment.id);
+  fetch('/delete-data', {method: 'POST', body: params});
+}
```
add new line

lorikeet-s commented on walkthroughs/week-3-server/examples/random-quotes/src/main/webapp/index.html:
```diff
@@ -10,7 +10,7 @@
       <h1>Random Quotes</h1>
       <p>Click here to see a random quote:</p>
-      <button onclick="getRandomQuote()">Fetch a quote!</button>
+      <button onclick="getRandomQuoteUsingArrowFunctions()">Fetch a quote!</button>
```
was this change supposed to be in here? it doesn't seem to relate to everything else.

lorikeet-s commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -14,19 +14,66 @@
+      comments.add(comment);
+    }
+
+    //Send the JSON as response
```
nit: comments should have a space after the // and end with a full stop.

lorikeet-s commented on portfolio/src/main/java/com/google/sps/data/Comment.java:
```diff
@@ -0,0 +1,31 @@
+  private final String commentText;
+  private final long timestamp;
+
+  public Comment(long id, String username, String commentText, long timestamp) {
```
This class and this function should have javadoc comments.
https://google.github.io/styleguide/javaguide.html#s7.3-javadoc-where-required

lorikeet-s replied:
Can you also add the block tags for the parameters: https://google.github.io/styleguide/javaguide.html#s7.1.3-javadoc-block-tags

helenatran replied:
Thanks @lorikeet-s, I just made the change. I am not sure whether this is how it was supposed to be done, so let me know if it needs to be modified again

lorikeet-s replied:
looks good things, for other functions you should have a return, but you don't normally add one for constructors I don't think.

lorikeet-s commented on portfolio/src/main/java/com/google/sps/data/Comment.java:
```diff
@@ -0,0 +1,31 @@
+    this.commentText = commentText;
+    this.timestamp = timestamp;
+  }
+}
```
add a new line

lorikeet-s reviewed (changes requested):
just some minor style issues.

helenatran committed:
Fix issues from GitHub review

helenatran committed:
Update .gitignore to remove '.idea'

helenatran commented:
Thanks @lorikeet-s! I made the requested changes, let me know whether there is anything else I can change.

helenatran committed:
Add block tags for parameters of Comment function
